TECHMANIA/Assets/Scripts/Components/Editor Scene/PatternPanel.cs

```diff
@@ -714,8 +741,60 @@ private void HandleKeyboardShortcuts()
                 maxPulse));
         }
     }
+    private void DragWorkSpace (Vector2 deltaPosition)
+    {
+        float outOfViewWidth = WorkspaceContentWidth -
+            workspaceViewport.rect.width;
+        float outOfViewHeight = WorkspaceContentHeight -
+            workspaceViewport.rect.height;
+        if (outOfViewWidth < 0f) outOfViewWidth = 0f;
+        if (outOfViewHeight < 0f) outOfViewHeight = 0f;
+
+        float horizontal =
+            workspaceScrollRect.horizontalNormalizedPosition *
+            outOfViewWidth;
+        horizontal -= deltaPosition.x / rootCanvas.localScale.x;
+        workspaceScrollRect.horizontalNormalizedPosition =
+            Mathf.Clamp01(horizontal / outOfViewWidth);
+
+        float vertical =
+            workspaceScrollRect.verticalNormalizedPosition *
+            outOfViewHeight;
+        vertical -= deltaPosition.y / rootCanvas.localScale.x;
+        workspaceScrollRect.verticalNormalizedPosition =
+            Mathf.Clamp01(vertical / outOfViewHeight);
+
+        SynchronizeScrollRects();
+    }
     #endregion
 
+    #region Touch
+    private void HandleTouchResize ()
+    {
+        Touch touchZero = Input.GetTouch(0);
+        Touch touchOne = Input.GetTouch(1);
+
+        Vector2 touchZeroPrevPos = touchZero.position - touchZero.deltaPosition;
```

Comment:
Since zoom is only horizontal, it may make more sense to only take into account the x coordinate from these positions and deltas.

Reply:
I think limit it to x diffence only will make it harder to use for small screen mobile users.